File: java/project/src/main/java/com/example/project/api/ReleaseAPI.java
package com.example.project.api;

import com.example.project.common.Paging;
import com.example.project.common.Result;
import com.example.project.dataobject.MainReleaseDO;
import com.example.project.dataobject.ReleaseDO;
import com.example.project.dataobject.ReleaseHistoryDO;
import com.example.project.model.MainRelease;
import com.example.project.model.Release;
import com.example.project.model.ReleaseHistory;
import com.example.project.service.ReleaseService;
import com.example.project.util.KeywordUtil;
import com.example.project.util.RedisOperationUtils.LockUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import javax.annotation.PostConstruct;
import java.util.List;
import java.util.Map;

/**
 * 与release相关的操作的api
 */
@Controller
@RequestMapping("/release")
public class ReleaseAPI {

    @Autowired
    private ReleaseService releaseService;

    @Autowired
    private RedisTemplate redisTemplate;

    @Autowired
    private LockUtil lockUtil;

    /**
     * 程序启动时进行初始化
     */
    @PostConstruct
    public void init() {
        releaseService.getNotAuditReleaseVersion();
    }

    /**
     * 启动后每隔五分钟进行一次感谢缓存的持久化
     */
    @Scheduled(cron = "* */5 *  * * ? ")
    public void releaseContributeThanksCount() {
        System.out.println("我开始持久化点赞数据啦");
        releaseService.releaseContributeThanksCount();
    }

    /**
     * 通过id获取release
     * @param id
     * @return
     */
    @GetMapping("/id/{id}")
    @ResponseBody
    public Result<Release> getReleaseById(@PathVariable("id") Long id) {
        Release release = releaseService.getReleaseById(id);
        if(release != null) {
            return Result.success(release);
        } else {
            return Result.error("Get release by id failed.");
        }
    }

    /**
     * 通过一组id获取一组release
     * @param ids
     * @return
     */
    @GetMapping("/ids")
    @ResponseBody
    public Result<List<Release>> getReleasesByIds(@RequestParam("ids")List<Long> ids) {
        List<Release> releases = releaseService.getReleasesByIds(ids);
        if(releases != null && releases.size() >0) {
            return Result.success(releases);
        } else {
            return Result.error("Get releases by ids failed.");
        }
    }

    /**
     * 修改release
     * @param releaseDO
     * @return
     */
    @PostMapping("/update")
    @ResponseBody
    public Result updateRelease(@RequestBody ReleaseDO releaseDO) {
        int i = releaseService.updateRelease(releaseDO);
        if(i==1) {
            return Result.success("Update release success!");
        } else {
            return Result.error("Update release failed!");
        }
    }

    /**
     * 添加release
     * @param releaseDO
     * @return
     */
    @PutMapping("/add")
    @ResponseBody
    public Result addRelease(@RequestBody ReleaseDO releaseDO) {
        Long id = releaseService.addRelease(releaseDO);
        if(id!=null) {
            return Result.success(id);
        } else {
            return Result.error("Add release failed");
        }
    }

    /**
     * release修改冲突检测
     * @param releaseId
     * @param userId
     * @return
     */
    @GetMapping("/startModify")
    @ResponseBody
    public Result startModify(@RequestParam("releaseId") Long releaseId, @RequestParam("userId") Long userId) {
        boolean isLock = lockUtil.isLock(releaseId, LockUtil.MODIFYING_RELEASE);
        if(isLock) {
            return Result.error("Start modify error");
        } else {
            return Result.success("Start modify success!");
        }
    }

    /**
     * 根据id获取mainRelease
     * @param mainId
     * @return
     */
    @GetMapping("/main/id/{mainId}")
    @ResponseBody
    public Result<MainRelease> getMainReleaseById(@PathVariable("mainId")Long mainId) {
        MainRelease mainRelease = releaseService.getMainReleaseById(mainId);
        if(mainRelease != null) {
            return Result.success(mainRelease);
        } else {
            return Result.error("Get main release by id failed!");
        }
    }

    /**
     * 根据关键词获取mainRelease
     * @param keyword
     * @param pageNum
     * @param pageSize
     * @return
     */
    @GetMapping("/main/search")
    @ResponseBody
    public Result<Paging<MainRelease>> getMainReleaseByKeyword(@RequestParam("keyword")String keyword, @RequestParam("pageNum") int pageNum, @RequestParam("pageSize") int pageSize) {
        Paging<MainRelease> mainReleases = releaseService.getMainReleaseByKeyword(keyword, pageNum, pageSize);
        if(mainReleases != null) {
            return Result.success(mainReleases);
        } else {
            return Result.error("Get main release by keyword failed!");
        }
    }

    /**
     * 添加mainRelease
     * @param mainReleaseDO
     * @return
     */
    @PutMapping("/main/add")
    @ResponseBody
    public Result addMainRelease(@RequestBody MainReleaseDO mainReleaseDO) {
        Long id = releaseService.addMainRelease(mainReleaseDO);
        if(id!=null) {
            return Result.success(id);
        } else {
            return Result.error("Add main release failed");
        }
    }

    /**
     * 获取一个release审核任务
     * @return
     */
    @GetMapping("/mission")
    @ResponseBody
    public Result dispatchMission() {
        ReleaseHistory releaseHistory = releaseService.dispatchMission();
        if(releaseHistory == null) {
            return Result.error("No mission now");
        } else {
            return Result.success(releaseHistory);
        }
    }

    /**
     * 添加releaseHistory
     * @param releaseHistoryDO
     * @return
     */
    @PutMapping("/version/add")
    @ResponseBody
    public Result addReleaseVersion(@RequestBody ReleaseHistoryDO releaseHistoryDO) {
        if(releaseHistoryDO.getIsNew()!=1) {
            boolean isLock = lockUtil.isLock(releaseHistoryDO.getReleaseId(), LockUtil.MODIFYING_RELEASE);
            if (isLock) {
                return Result.error("Add release version failed!");
            }
        }
        int addNum = releaseService.addReleaseVersion(releaseHistoryDO);
        if(addNum <= 0) {
            return Result.error("Add release version failed!");
        } else {
            if(releaseHistoryDO.getIsNew()!=1) {
                lockUtil.lock(releaseHistoryDO.getReleaseId(), releaseHistoryDO.getUserId(), LockUtil.MODIFYING_RELEASE);
            }
            return Result.success("Add release version succeed!");
        }
    }

    /**
     * 更新releaseHistory
     * @param releaseHistoryDO
     * @return
     */
    @PostMapping("/version/update")
    @ResponseBody
    public Result updateReleaseVersion(@RequestBody ReleaseHistoryDO releaseHistoryDO) {
        int i = releaseService.updateReleaseVersion(releaseHistoryDO);
        if(i==1) {
            return Result.success("Audit release success!");
        } else {
            return Result.error("Audit release failed!");
        }
    }

    /**
     * 根据id获得release历史记录
     * @param releaseId
     * @return
     */
    @GetMapping("/version/all")
    @ResponseBody
    public Result getAllReleaseVersion(@RequestParam("releaseId") Long releaseId, @RequestParam("userId")Long userId) {
        List<ReleaseHistory> releaseHistories = releaseService.getAllReleaseVersion(releaseId, userId);
        if(releaseHistories!=null&&releaseHistories.size()>0) {
            return Result.success(releaseHistories);
        } else {
            return Result.error("Get all release version failed");
        }
    }

    /**
     * 提交release审核
     * @param auditMap
     * @return
     */
    @PostMapping("/audit")
    @ResponseBody
    public Result auditRelease(@RequestBody Map auditMap) {
        boolean result = releaseService.auditRelease(auditMap);
        if(result) {
            return Result.success("Audit release success!");
        } else {
            return Result.error("Audit release error!");
        }
    }

    /**
     * 根据artistId获取release总数
     * @param artistId
     * @return
     */
    @GetMapping("/artist/count/{artistId}")
    @ResponseBody
    public Result<Map> getArtistReleaseCount(@PathVariable("artistId") Long artistId) {
        return Result.success(releaseService.getArtistReleaseCount(artistId));
    }

    /**
     * 根据artistId获取release
     * @param artistId
     * @param pageNum
     * @param pageSize
     * @return
     */
    @GetMapping("/artist/release/{artistId}")
    @ResponseBody
    public Result<Paging<Release>> getArtistRelease(@PathVariable("artistId") Long artistId, @RequestParam("pageNum") int pageNum, @RequestParam("pageSize") int pageSize) {
        Paging<Release> res = releaseService.getArtistRelease(artistId, pageNum, pageSize);
        if(res.getData() != null) {
            return Result.success(res);
        } else {
            return Result.error("Get artist release failed.");
        }
    }

    /**
     * 根据artistId获取featured release
     * @param artistId
     * @param pageNum
     * @param pageSize
     * @return
     */
    @GetMapping("/artist/featured/{artistId}")
    @ResponseBody
    public Result<Paging<Release>> getArtistFeatured(@PathVariable("artistId") Long artistId, @RequestParam("pageNum") int pageNum, @RequestParam("pageSize") int pageSize) {
        Paging<Release> res = releaseService.getArtistFeatured(artistId, pageNum, pageSize);
        if(res.getData() != null) {
            return Result.success(res);
        } else {
            return Result.error("Get artist featured failed.");
        }
    }

    /**
     * 根据artistId获取behind release
     * @param artistId
     * @param pageNum
     * @param pageSize
     * @return
     */
    @GetMapping("/artist/behind/{artistId}")
    @ResponseBody
    public Result<Paging<Release>> getArtistBehind(@PathVariable("artistId") Long artistId, @RequestParam("pageNum") int pageNum, @RequestParam("pageSize") int pageSize) {
        Paging<Release> res = releaseService.getArtistBehind(artistId, pageNum, pageSize);
        if(res.getData() != null) {
            return Result.success(res);
        } else {
            return Result.error("Get artist behind failed!");
        }
    }

    /**
     * 根据labelId获取release
     * @param labelId
     * @param pageNum
     * @param pageSize
     * @return
     */
    @GetMapping("/label/{labelId}")
    @ResponseBody
    public Result<Paging<Release>> getLabelRelease(@PathVariable("labelId") Long labelId, @RequestParam("pageNum") int pageNum, @RequestParam("pageSize") int pageSize) {
        Paging<Release> result = releaseService.getLabelRelease(labelId, pageNum, pageSize);
        if(result.getData() != null) {
            return Result.success(result);
        } else {
            return Result.error("Get label releases failed!");
        }

    }
}

File: js/src/api/releaseAPI.ts
import request from "@/api/request"

import type { releaseData } from "@/api/dataInterfaces"

import sleep from "@/common/sleep"

export const getReleaseById = async (id: number) => {
  const result = await request(`/release/id/${id}`)
  if (result.error) {
    return result.error
  } else if (result.isSuccess) {
    return result.data
  } else {
    return result.message
  }
}

export const getReleasesByIds = async (ids: number[]) => {
  const result = await request(`/release/ids`, ids)
  if (result.error) {
    return result.error
  } else if (result.isSuccess) {
    return result.data
  } else {
    return result.message
  }
}

export const getReleaseVerions = async (mainId: number) => {
  const result = await request(`/release/version/${mainId}`)
  if (result.isSuccess) {
    return result.data
  } else {
    return null
  }
}

export const getArtistReleaseCount = async (artistId: number) => {
  const result = await request(`/release/artist/count/${artistId}`)
  if (result.isSuccess) {
    return result.data
  } else {
    return null
  }
}

export const getArtistRelease = async (artistId: number, pageNum: number, pageSize: number) => {
  const result = await request(`/release/artist/release/${artistId}`, { pageNum, pageSize })
  if (result.isSuccess) {
    return result.data.data
  } else {
    return null
  }
}

export const getArtistFeatured = async (artistId: number, pageNum: number, pageSize: number) => {
  const result = await request(`/release/artist/featured/${artistId}`, { pageNum, pageSize })
  if (result.isSuccess) {
    return result.data.data
  } else {
    return null
  }
}

export const getArtistBehind = async (artistId: number, pageNum: number, pageSize: number) => {
  const result = await request(`/release/artist/behind/${artistId}`, { pageNum, pageSize })
  if (result.error) {
    return result.error
  } else if (result.isSuccess) {
    return result.data.data
  } else {
    return result.message
  }
}

export const getLabelRelease = async (labelId: number, pageNum: number, pageSize: number) => {
  const result = await request(`/release/label/${labelId}`, { pageNum, pageSize })
  if (result.error) {
    return result.error
  } else if (result.isSuccess) {
    return result.data
  } else {
    return result.message
  }
}

export const getMainReleaseById = async (mainReleaseId: number) => {
  const result = await request(`/release/main/id/${mainReleaseId}`)
  if (result.error) {
    return result.error
  } else if (result.isSuccess) {
    return result.data
  } else {
    return result.message
  }
}

export const getMainReleaseByKeyword = async (keyword: string, pageNum: number, pageSize: number) => {
  const result = await request(`/release/main/search`, { keyword, pageNum, pageSize })
  if (result.error) {
    return result.error
  } else if (result.isSuccess) {
    return result.data
  } else {
    return result.message
  }
}

export const addReleaseVersion = async (newVersion: any) => {
  const str = JSON.stringify(newVersion)
  const result = await request(`/release/version/add`, str, "PUT", {
    token: localStorage.getItem("accessToken") || ""
  })
  if (result.error) {
    return {
      message: result.error,
      isSuccess: false
    }
  } else if (result.isSuccess) {
    return {
      message: "提交成功！",
      isSuccess: true
    }
  } else {
    return {
      message: result.message,
      isSuccess: false
    }
  }
}

export const getMission = async () => {
  // const result = {
  //     auditorId: null,
  //     change: null,
  //     gmtCreated: "2023-03-30 13:15:48",
  //     gmtModified: "2023-03-30 13:15:48",
  //     id: 3,
  //     info: "{\"form\": \"3\", \"genre\": [\"Electronic\"], \"style\": [\"Dubstep\", \"Riddim Dubstep\"], \"title\": \"111\", \"artist\": [{\"id\": 3619817, \"name\": \"Barely Alive\"}], \"format\": \"4\", \"isBase\": false, \"mainId\": 1061731, \"country\": \"US\", \"credits\": null, \"picture\": null, \"tracklist\": [{\"num\": 0, \"open\": false, \"title\": \"1\", \"artist\": [], \"duration\": \"\", \"position\": \"1\", \"extraartists\": []}], \"supplement\": \"11\", \"mainRelease\": {\"id\": 1061731, \"isNew\": false, \"title\": \"We Are Barely Alive\"}, \"releaseDate\": \"2023-03-03\", \"releaseLabel\": [{\"id\": 999909, \"name\": \"Disciple (3)\"}]}",
  //     isPass: "wait",
  //     new: true,
  //     reference: "1",
  //     release: {
  //       artist: null,
  //       country: null,
  //       credits: null,
  //       form: null,
  //       format: null,
  //       genre: null,
  //       gmtCreated: null,
  //       gmtModified: null,
  //       id: null,
  //       isBase: false,
  //       mainRelease: null,
  //       otherVersion: null,
  //       picture: null,
  //       releaseDate: null,
  //       releaseLabel: null,
  //       style: null,
  //       supplement: null,
  //       title: null,
  //       tracklist: null,
  //     },
  //     title: "111",
  //     userId: 2,
  //     userName: "2",
  // }

  // await sleep(1000)

  // return result
  
  const result = await request(`/release/mission`, {}, "get", {
    token: localStorage.getItem("accessToken") || ""
  })
  if(result.error) {
    return result.error
  } else if(result.isSuccess) {
    return result.data
  } else {
    return result.message
  }
}

export const updateReleaseVersion = async (releaseHistory: any) => {
  const str = JSON.stringify(releaseHistory)
  const result = await request(`/release/version/update`, str, "POST", {
    token: localStorage.getItem("accessToken") || ""
  })
  if (result.error) {
    return {
      message: result.error,
      isSuccess: false
    }
  } else if (result.isSuccess) {
    return {
      message: result.data,
      isSuccess: true
    }
  } else {
    return {
      message: result.message,
      isSuccess: false
    }
  }
}

export const updateRelease = async (release: any) => {
  const str = JSON.stringify(release)
  const result = await request(`/release/update`, str, "POST", {
    token: localStorage.getItem("accessToken") || ""
  })
  if (result.error) {
    return {
      message: result.error,
      isSuccess: false
    }
  } else if (result.isSuccess) {
    return {
      message: result.data,
      isSuccess: true
    }
  } else {
    return {
      message: result.message,
      isSuccess: false
    }
  }
}

export const addRelease = async (release: any) => {
  const str = JSON.stringify(release)
  const result = await request(`/release/add`, str, "PUT", {
    token: localStorage.getItem("accessToken") || ""
  })
  if (result.error) {
    return {
      message: result.error,
      isSuccess: false
    }
  } else if (result.isSuccess) {
    return {
      data: result.data,
      isSuccess: true
    }
  } else {
    return {
      message: result.message,
      isSuccess: false
    }
  }
}

export const getAllReleaseVersion = async (releaseId: number, userId: number)=> {
  const result = await request(`/release/version/all`, {releaseId, userId})
  if(result.error) {
    return result.error
  } else if(result.isSuccess) {
    return result.data
  } else {
    return result.message
  }
}

export const addMainRelease = async (mainRelease: any)=> {
  const str = JSON.stringify(mainRelease)
  const result = await request(`/release/main/add`, str, "PUT", {
    token: localStorage.getItem("accessToken") || ""
  })
  if (result.error) {
    return {
      message: result.error,
      isSuccess: false
    }
  } else if (result.isSuccess) {
    return {
      data: result.data,
      isSuccess: true
    }
  } else {
    return {
      message: result.message,
      isSuccess: false
    }
  }
}

export const startModify = async(releaseId: number, userId: number)=> {
  return await request(`/release/startModify`, {releaseId, userId}, "get", {
    token: localStorage.getItem("accessToken") || ""
  })
}

export const auditRelease = async(auditMap: any)=> {
  const str = JSON.stringify(auditMap)
  const result = await request(`/release/audit`, str, "post", {
    token: localStorage.getItem("accessToken") || ""
  })
  return result.isSuccess
}